File: 1306-jump-game-iii/1306-jump-game-iii.java
class Solution {
    public boolean canReach(int[] arr, int start) {
        int n = arr.length;
        int[] vis = new int[n];
        Queue<Integer> q = new LinkedList<>();
        q.offer(start);
        vis[start] = 1;
        while(!q.isEmpty()){
            int node = q.peek();
            q.poll();
            if(arr[node] == 0) return true;

            int left = node - arr[node];
            int right = node + arr[node];

            if(left >= 0 && vis[left] == 0){
                q.offer(left);
                vis[left] = 1;
            }
            if(right < n && vis[right] == 0){
                q.offer(right);
                vis[right] = 1;
            }
        }
        return false;
    }
}

File: 1306-jump-game-iii/1306-jump-game-iii.cpp
class Solution {
public:
    bool canReach(vector<int>& arr, int start) {
        int n = arr.size();
        queue<int> q;
        q.push(start);
        if(arr[start] == 0) return true;
        vector<int> vis(n , 0);
        vis[start] = 1;
        while(!q.empty() && q.size() < n){
            int node = q.front();q.pop();
            if(arr[node] == 0) return true;
            if(node - arr[node] >= 0  && !vis[node - arr[node]]){ 
                q.push(node - arr[node]);
                vis[node - arr[node]] = 1;
            }
            if(node + arr[node] < n && !vis[node + arr[node]]) {
                q.push(node + arr[node]);
                vis[node + arr[node]] = 1;
            }
        }
        return false;
    }
};